File: src/Trulla/Ast.fs
namespace Trulla.Core

open TheBlunt

open Trulla.Core
open Trulla.Core.Utils

type TVar =
    | Root
    | TVar of int

type BindingContext = Map<string, TVar>

type TVal<'a> =
    { 
        range: Range
        tvar: TVar
        bindingContext: BindingContext
        value: 'a 
    }
    override this.ToString() = sprintf "(%s)%s" (this.range.ToString()) (this.value.ToString())

type TExp =
    | Text of string
    | Hole of TVal<MemberExp>
    | For of ident: TVal<string> * exp: TVal<MemberExp> * sep: PVal<string option> * body: TExp list
    | If of cond: TVal<MemberExp> * body: TExp list
    | Else of cond: TVal<MemberExp> * body: TExp list

and Body = BindingContext * TExp list

and MemberExp =
    | AccessExp of {| instanceExp: TVal<MemberExp>; memberName: string |}
    | IdentExp of string

type Ast =
    {
        tree: TExp list
        tvarToMemberExp: Map<TVar, MemberExp>
    }

[<RequireQualifiedAccess>]
module Ast =

    [<RequireQualifiedAccess>]
    type private Scope =
        | IfOrElseScope of cond: TVal<MemberExp>
        | Other
    
    let createTVal range tvar bindingContext value =
        { range = range; tvar = tvar; bindingContext = bindingContext; value = value }
    
    // TODO: meaningful error messages + location
    // TODO: Don't throw; return TemplateError results
    let buildTree (tokens: PVal<Token> list) : Result<_,_> =
        // TODO: This is far from what I thing it could be in terms of understandable code
        
        let mutable tvarToMemberExp = Map.empty

        let newTVar =
            let mutable x = -1
            fun () ->
                x <- x + 1
                TVar x

        let buildMemberExp bindingContext pexp =
            let rec ofPExpZero (pexp: PVal<MemberToken>) =
                let newTVal tvar value = createTVal pexp.range tvar bindingContext value
                let newTValAndAdd exp =
                    let tvar = newTVar ()
                    do tvarToMemberExp <- tvarToMemberExp |> Map.add tvar exp
                    newTVal tvar exp
                match pexp.result with
                | IdentToken ident ->
                    newTValAndAdd (IdentExp ident)
                | AccessToken accExp ->
                    let accExp = {| instanceExp = ofPExpZero accExp.instanceExp; memberName = accExp.memberName |}
                    newTValAndAdd (AccessExp accExp)
            ofPExpZero pexp

        let mutable currTokIdx = 0
        let mutable openScopeStack = []
        let mutable elseBlockOpen = false
        let rec toTree (bindingContext: BindingContext) : Result<_,_> =
            let mutable scopeClosed = false
            let mutable revTree = []
            let mutable errors = []
            
            let addToken x = do revTree <- x :: revTree

            do while not scopeClosed && currTokIdx < tokens.Length do
                let token = tokens.[currTokIdx]
                currTokIdx <- currTokIdx + 1

                let addError message =
                    do errors <-
                        [
                            yield! errors
                            yield 
                                {
                                    range = token.range
                                    message = message
                                }
                        ]
                let addErrors newErrors =
                    do errors <- [ yield! errors; yield! newErrors ]

                match token.result with
                | Token.Text x -> 
                    let x = Text x
                    do addToken x
                | Token.Hole x -> 
                    let x = Hole (buildMemberExp bindingContext x)
                    do addToken x
                | Token.For (ident, acc, sep) ->
                    let accExp = buildMemberExp bindingContext acc
                    let tvarIdent = newTVar()
                    do openScopeStack <- Scope.Other :: openScopeStack
                    match toTree (Map.add ident.result tvarIdent bindingContext) with
                    | Ok subTree -> 
                        let x = 
                            (
                                createTVal ident.range tvarIdent bindingContext ident.result,
                                accExp,
                                sep,
                                subTree
                            )
                            |> For
                        do addToken x
                    | Error subErrors ->
                        do addErrors subErrors
                | Token.If acc ->
                    let cond = buildMemberExp bindingContext acc
                    do openScopeStack <- (Scope.IfOrElseScope cond) :: openScopeStack
                    match toTree bindingContext with
                    | Ok subTree -> do addToken (If (cond, subTree))
                    | Error subErrors -> do addErrors subErrors
                | Token.Else ->
                    if elseBlockOpen then
                        do elseBlockOpen <- false
                        match openScopeStack with
                        | (Scope.IfOrElseScope cond) :: _ ->
                            match toTree bindingContext with
                            | Ok subTree -> do addToken (Else (cond, subTree))
                            | Error subErrors -> do errors <- [ yield! errors; yield! subErrors ]
                        | _ ->
                            do addError "An else needs an if."
                    else
                        do currTokIdx <- currTokIdx - 1
                        do elseBlockOpen <- true
                        do scopeClosed <- true
                | Token.End ->
                    match openScopeStack with
                    | [] -> 
                        do addError "Closing a scope is not possible without having a scope open." 
                    | _::xs -> 
                        do openScopeStack <- xs
                        do scopeClosed <- true

            match errors with
            | [] -> revTree |> List.rev |> Ok
            | _ -> Error errors

        match toTree Map.empty with
        | Ok tree ->
            if openScopeStack.Length > 0 then
                // TODO: Range.zero is wrong
                { range = Range.zero; message = "TODO: Unclosed scope detected." }
                |> List.singleton
                |> Error
            else
                Ok { tree = tree; tvarToMemberExp = tvarToMemberExp }
        | Error errors -> Error errors

File: src/NodeCli/lib/Common/CoreFiles/Ast.js
import { Record, toString, Union } from "../../fable_modules/fable-library.4.3.0/Types.js";
import { anonRecord_type, list_type, option_type, record_type, class_type, string_type, union_type, int32_type } from "../../fable_modules/fable-library.4.3.0/Reflection.js";
import { printf, toText } from "../../fable_modules/fable-library.4.3.0/String.js";
import { RangeModule_zero, PVal$1_$reflection, Range$_$reflection } from "./TheBlunt.js";
import { add, empty } from "../../fable_modules/fable-library.4.3.0/Map.js";
import { comparePrimitives, compare } from "../../fable_modules/fable-library.4.3.0/Util.js";
import { singleton as singleton_1, reverse, tail, head, isEmpty, item, length, cons, empty as empty_1 } from "../../fable_modules/fable-library.4.3.0/List.js";
import { singleton, append, delay, toList } from "../../fable_modules/fable-library.4.3.0/Seq.js";
import { TrullaError } from "./Core.js";
import { FSharpResult$2 } from "../../fable_modules/fable-library.4.3.0/Choice.js";

export class TVar extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Root", "TVar"];
    }
}

export function TVar_$reflection() {
    return union_type("Trulla.Core.TVar", [], TVar, () => [[], [["Item", int32_type]]]);
}

export class TVal$1 extends Record {
    constructor(range, tvar, bindingContext, value) {
        super();
        this.range = range;
        this.tvar = tvar;
        this.bindingContext = bindingContext;
        this.value = value;
    }
    toString() {
        const this$ = this;
        const arg = toString(this$.range);
        let arg_1;
        let copyOfStruct = this$.value;
        arg_1 = toString(copyOfStruct);
        return toText(printf("(%s)%s"))(arg)(arg_1);
    }
}

export function TVal$1_$reflection(gen0) {
    return record_type("Trulla.Core.TVal`1", [gen0], TVal$1, () => [["range", Range$_$reflection()], ["tvar", TVar_$reflection()], ["bindingContext", class_type("Microsoft.FSharp.Collections.FSharpMap`2", [string_type, TVar_$reflection()])], ["value", gen0]]);
}

export class TExp extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Text", "Hole", "For", "If", "Else"];
    }
}

export function TExp_$reflection() {
    return union_type("Trulla.Core.TExp", [], TExp, () => [[["Item", string_type]], [["Item", TVal$1_$reflection(MemberExp_$reflection())]], [["ident", TVal$1_$reflection(string_type)], ["exp", TVal$1_$reflection(MemberExp_$reflection())], ["sep", PVal$1_$reflection(option_type(string_type))], ["body", list_type(TExp_$reflection())]], [["cond", TVal$1_$reflection(MemberExp_$reflection())], ["body", list_type(TExp_$reflection())]], [["cond", TVal$1_$reflection(MemberExp_$reflection())], ["body", list_type(TExp_$reflection())]]]);
}

export class MemberExp extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["AccessExp", "IdentExp"];
    }
}

export function MemberExp_$reflection() {
    return union_type("Trulla.Core.MemberExp", [], MemberExp, () => [[["Item", anonRecord_type(["instanceExp", TVal$1_$reflection(MemberExp_$reflection())], ["memberName", string_type])]], [["Item", string_type]]]);
}

export class Ast extends Record {
    constructor(tree, tvarToMemberExp) {
        super();
        this.tree = tree;
        this.tvarToMemberExp = tvarToMemberExp;
    }
}

export function Ast_$reflection() {
    return record_type("Trulla.Core.Ast", [], Ast, () => [["tree", list_type(TExp_$reflection())], ["tvarToMemberExp", class_type("Microsoft.FSharp.Collections.FSharpMap`2", [TVar_$reflection(), MemberExp_$reflection()])]]);
}

class AstModule_Scope extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["IfOrElseScope", "Other"];
    }
}

function AstModule_Scope_$reflection() {
    return union_type("Trulla.Core.AstModule.Scope", [], AstModule_Scope, () => [[["cond", TVal$1_$reflection(MemberExp_$reflection())]], []]);
}

export function AstModule_createTVal(range, tvar, bindingContext, value) {
    return new TVal$1(range, tvar, bindingContext, value);
}

export function AstModule_buildTree(tokens) {
    let tvarToMemberExp = empty({
        Compare: compare,
    });
    let newTVar;
    let x_1 = -1;
    newTVar = (() => {
        x_1 = ((x_1 + 1) | 0);
        return new TVar(1, [x_1]);
    });
    const buildMemberExp = (bindingContext, pexp) => {
        const ofPExpZero = (pexp_1) => {
            const newTValAndAdd = (exp) => {
                const tvar_1 = newTVar();
                tvarToMemberExp = add(tvar_1, exp, tvarToMemberExp);
                return AstModule_createTVal(pexp_1.range, tvar_1, bindingContext, exp);
            };
            const matchValue = pexp_1.result;
            if (matchValue.tag === 0) {
                const accExp = matchValue.fields[0];
                return newTValAndAdd(new MemberExp(0, [{
                    instanceExp: ofPExpZero(accExp.instanceExp),
                    memberName: accExp.memberName,
                }]));
            }
            else {
                return newTValAndAdd(new MemberExp(1, [matchValue.fields[0]]));
            }
        };
        return ofPExpZero(pexp);
    };
    let currTokIdx = 0;
    let openScopeStack = empty_1();
    let elseBlockOpen = false;
    const toTree = (bindingContext_1) => {
        let tupledArg;
        let scopeClosed = false;
        let revTree = empty_1();
        let errors = empty_1();
        const addToken = (x_2) => {
            revTree = cons(x_2, revTree);
        };
        while (!scopeClosed && (currTokIdx < length(tokens))) {
            const token = item(currTokIdx, tokens);
            currTokIdx = ((currTokIdx + 1) | 0);
            const addError = (message) => {
                errors = toList(delay(() => append(errors, delay(() => singleton(new TrullaError(token.range, message))))));
            };
            const addErrors = (newErrors) => {
                errors = toList(delay(() => append(errors, delay(() => newErrors))));
            };
            const matchValue_1 = token.result;
            switch (matchValue_1.tag) {
                case 1: {
                    addToken(new TExp(1, [buildMemberExp(bindingContext_1, matchValue_1.fields[0])]));
                    break;
                }
                case 2: {
                    const ident_1 = matchValue_1.fields[0];
                    const accExp_2 = buildMemberExp(bindingContext_1, matchValue_1.fields[1]);
                    const tvarIdent = newTVar();
                    openScopeStack = cons(new AstModule_Scope(1, []), openScopeStack);
                    const matchValue_2 = toTree(add(ident_1.result, tvarIdent, bindingContext_1));
                    if (matchValue_2.tag === 1) {
                        addErrors(matchValue_2.fields[0]);
                    }
                    else {
                        addToken((tupledArg = [AstModule_createTVal(ident_1.range, tvarIdent, bindingContext_1, ident_1.result), accExp_2, matchValue_1.fields[2], matchValue_2.fields[0]], new TExp(2, [tupledArg[0], tupledArg[1], tupledArg[2], tupledArg[3]])));
                    }
                    break;
                }
                case 3: {
                    const cond = buildMemberExp(bindingContext_1, matchValue_1.fields[0]);
                    openScopeStack = cons(new AstModule_Scope(0, [cond]), openScopeStack);
                    const matchValue_3 = toTree(bindingContext_1);
                    if (matchValue_3.tag === 1) {
                        addErrors(matchValue_3.fields[0]);
                    }
                    else {
                        addToken(new TExp(3, [cond, matchValue_3.fields[0]]));
                    }
                    break;
                }
                case 4: {
                    if (elseBlockOpen) {
                        elseBlockOpen = false;
                        let matchResult, cond_1;
                        if (!isEmpty(openScopeStack)) {
                            if (head(openScopeStack).tag === 0) {
                                matchResult = 0;
                                cond_1 = head(openScopeStack).fields[0];
                            }
                            else {
                                matchResult = 1;
                            }
                        }
                        else {
                            matchResult = 1;
                        }
                        switch (matchResult) {
                            case 0: {
                                const matchValue_4 = toTree(bindingContext_1);
                                if (matchValue_4.tag === 1) {
                                    errors = toList(delay(() => append(errors, delay(() => matchValue_4.fields[0]))));
                                }
                                else {
                                    addToken(new TExp(4, [cond_1, matchValue_4.fields[0]]));
                                }
                                break;
                            }
                            case 1: {
                                addError("An else needs an if.");
                                break;
                            }
                        }
                    }
                    else {
                        currTokIdx = ((currTokIdx - 1) | 0);
                        elseBlockOpen = true;
                        scopeClosed = true;
                    }
                    break;
                }
                case 5: {
                    if (!isEmpty(openScopeStack)) {
                        const xs = tail(openScopeStack);
                        openScopeStack = xs;
                        scopeClosed = true;
                    }
                    else {
                        addError("Closing a scope is not possible without having a scope open.");
                    }
                    break;
                }
                default:
                    addToken(new TExp(0, [matchValue_1.fields[0]]));
            }
        }
        if (isEmpty(errors)) {
            return new FSharpResult$2(0, [reverse(revTree)]);
        }
        else {
            return new FSharpResult$2(1, [errors]);
        }
    };
    const matchValue_5 = toTree(empty({
        Compare: comparePrimitives,
    }));
    if (matchValue_5.tag === 1) {
        return new FSharpResult$2(1, [matchValue_5.fields[0]]);
    }
    else if (length(openScopeStack) > 0) {
        return new FSharpResult$2(1, [singleton_1(new TrullaError(RangeModule_zero, "TODO: Unclosed scope detected."))]);
    }
    else {
        return new FSharpResult$2(0, [new Ast(matchValue_5.fields[0], tvarToMemberExp)]);
    }
}

